src/Microsoft.Azure.WebJobs.Host/JobHost.cs

```diff
@@ -329,37 +335,92 @@ private static IFunctionDefinition ResolveFunctionDefinition(MethodInfo method,
             return function;
         }
 
-        private async Task<JobHostContext> CreateContextAndLogHostStartedAsync(CancellationToken cancellationToken)
+        private void ThrowIfDisposed()
         {
-            JobHostContext context = await _config.CreateAndLogHostStartedAsync(this, _shutdownTokenSource.Token, cancellationToken);
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(null);
+            }
+        }
 
-            lock (_contextLock)
+        // If multiple threads call this, only one should do the init. The rest should wait
+        // When this task is signalled, _context is initialized. 
+        private Task EnsureHostStartedAsync(CancellationToken cancellationToken)
+        {
+            if (_context != null)
             {
-                if (_context == null)
+                return Task.CompletedTask;
+            }
+
+            TaskCompletionSource<bool> tsc = null;
+
+            lock (_stopTaskLock)
```

Comment:
This class should only have 1 lock.  I'll rename the field.  
And the logic before was pretty convoluted - I remove 4 state fields 

---
In reply to: [126255344](https://github.com/Azure/azure-webjobs-sdk/pull/1202#discussion_r126255344) [](ancestors = 126255344)